Simplify TS numbers drawings, fix bug related to refresh & local thermal generation

guilpier-code opened the pull request:
We intend to simplify and clarify the TS number management (specially source file **timeseries-numbers.cpp**).
For this we did several things : 
- [x] For generated TS, separate resizing from generation, and call the resizing before drawing TS number, so that we have the definitive TS sizes before we draw the TS numbers : this is the **main leverage** for TS number management simplifications and clarification.
- [x] At construction, the 1-d matrix **timeseriesNumbers** is given zeros. This allows many simplifications and code reductions.
- [ ] For productions (only for **Load**, **Wind** and **Solar**) that generate null TS (due to the configuration of their TS generator [see prepro capacity]), raise a warning somewhere for user to know.

guilpier-code committed:
Simplify TS numbers draws : make data members private

guilpier-code committed:
Simplify TS numbers draws : make nbHoursADay a constant unsigned int for a whole file

guilpier-code committed:
Simplify TS numbers draws : mainly extracting generated TS resizing from TS generation and move it before TS number drawings

guilpier-code committed:
Simplify TS numbers draws : first simplification on TS numbers drawings

guilpier-code committed:
Simplify TS numbers draws : also moving hydro generated TS before TS numbers drawing

guilpier-code committed:
Simplify TS numbers draws : simplifications on TS numbers drawings about hydro

guilpier-code committed:
Simplify TS numbers draws : also moving thermal generated TS before TS numbers drawing + simplification on TS number drawing

guilpier-code committed:
Simplify TS numbers draws : fix TS number unit tests

sylvlecl commented on src/solver/ts-generator/generator.hxx:
```diff
@@ -29,12 +29,52 @@
-{
-namespace TSGenerator
+ 
+inline void ResizeGeneratedTimeSeries(Data::AreaList& areas, Data::Parameters& params)
```
It does not seem templated --> should go in a .cpp

flomnes replied:
And `inline` can be removed, too

guilpier-code replied:
Done

sylvlecl commented on src/solver/ts-generator/generator.hxx:
```diff
@@ -29,12 +29,52 @@
+ 
+inline void ResizeGeneratedTimeSeries(Data::AreaList& areas, Data::Parameters& params)
 {
+    for (auto i = areas.begin(); i != areas.end(); ++i)
```
We should use range-based for loop, no need for iterator.

guilpier-code replied:
Done : for this, I used **each()**

sylvlecl commented on src/solver/ts-generator/generator.hxx:
```diff
@@ -29,12 +29,52 @@
+
+        // Thermal
+        auto end = area.thermal.list.mapping.end();
+        for (auto it = area.thermal.list.mapping.begin(); it != end; ++it)
```
we should use range-based for loop, no need for iterator

guilpier-code replied:
I can't see how to do it. If you see, please show an example.

guilpier-code replied:
Thanks to @JasonMarechal25 who suggested a correct solution (see code)

sylvlecl reviewed (commented):
A few minor comments

guilpier-code committed:
Simplify TS numbers draws : use timeSeriesToRefres instead of timeSeriesToGenerate to stay consistent with other part of the code.

But whether we should use one or the other in the code requires thinking, and possibly change

guilpier-code committed:
[skip ci] Simplify TS numbers draws : remove useless argument

guilpier-code committed:
Simplify TS numbers draws : removing knowledge of TS acquisition mode from intra-modal check

guilpier-code committed:
Simplify TS numbers draws : removing knowledge of TS acquisition mode from intra-modal check (part 2)

guilpier-code committed:
Simplify TS numbers draws : removing knowledge of TS acquisition mode from intra-modal check (part 3)

guilpier-code committed:
[skip ci] Simplify TS numbers draws : removing knowledge of TS acquisition mode from intra-modal check (part 4)

guilpier-code committed:
Simplify TS numbers draws : remove boolean (is intra modal ?) from intra modal consistency checker

flomnes commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -286,24 +252,17 @@ class IntraModalConsistencyChecker
-
+    
+    if (not checkTSconsistency())
+        return false;
```
`return checkTSconsistency();`

guilpier-code replied:
Apparently, the previous comment is not valid anymore : the code changed since then

flomnes commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -364,19 +322,20 @@ bool checkIntraModalConsistency(array<uint, timeSeriesCount>& nbTimeseriesByMode
+            IntraModalConsistencyChecker intraModalchecker(tsKind, tsRetriever, study);
+            if (!intraModalchecker.check())
+                return false;
+            nbTimeseriesByMode[indexTS] = intraModalchecker.getTimeSeriesNumber();
```
Careful, `nbTimeseriesByMode[indexTS]` isn't filled if `!intraModalchecker.check()` is `false`. Is this intended ?

guilpier-code replied:
This part hasn't changed, it's still like it used to be.
But some changes should occur here at some point.

guilpier-code committed:
Simplify TS numbers draws : fix results instability

guilpier-code committed:
Simplify TS numbers draws : remove check(), not needed anymore

guilpier-code committed:
Simplify TS numbers draws : simplify intra-modal check a bit

guilpier-code committed:
Simplify TS numbers draws : simplify intra-modal check a bit

guilpier-code committed:
Simplify TS numbers draws : remove study as TS retriever classes

guilpier-code committed:
[skip ci] Simplify TS numbers draws : make parent class's destructor virtual

guilpier-code committed:
Simplify TS numbers draws : remove study as a function argument

guilpier-code committed:
Simplify TS numbers draws : small simplification

guilpier-code committed:
Simplify TS numbers draws : test : removing empty TS management. Does it fail ?

guilpier-code committed:
Simplify TS numbers draws : resize mingen TS when hydro TS are generated

Behavior before this commit :
   When hydro TS are generated, mingen are not generated : it is sized with 1 column.
   So (given an area and a year) a TS number for ror and storage does not apply to mingen in this case.
   It causes extra complexity in class TimeSeries to handle this.
   Note that, still in this case (hydro TS are generated), the only mingen TS is filled with zeros.

Behavior after this commit :
   We remove the complexity in class TimeSeries by doing the following :
   If hydro TS are generated, mingen TS matrix is resized to the number of TS to be generated and filled with zeros.
   So we still have a mingen of zeros.

guilpier-code committed:
Simplify TS numbers draws : move permission to have 1 TS intra-modal where TS numbers are generated

2 commits ago, we removed ts numbers handling from class TimeSeries.
But it has side effect regarding results in case the energy is intra-modal : in this mode, having 1 TS only in an area is allowed.
The consequences of this allowance was handled by class TimeSeries. It had to be handled somewhere.
We move this code where TS numbers are generated, more specifically where intra-modal is handled.

flomnes commented on src/libs/antares/study/parts/hydro/series.cpp:
```diff
@@ -236,10 +236,11 @@ void DataSeriesHydro::reset()
-    storage.resize(width, DAYS_PER_YEAR);
+    ror.reset(width, HOURS_PER_YEAR);
+    storage.reset(width, DAYS_PER_YEAR);
+    mingen.reset(width, HOURS_PER_YEAR);
```
This will discard user input in some cases.

guilpier-code replied:
You're right : 
This function is called only when hydro TS are generated.

In this case, for **ror** and **storage** TS, we don't do anything new : TS must be resized to the number of TS the user gave in the TS management pane (or in the **input** equivalent on disk). The resized matrix is then filled when generating the TS.

**mingen** is not generated though, and the TS given by user (from files) must be kept intact.
So **mingen** TS matrix is resized to the generation number of TS (in TS management pane, hydro part) : 
- if new size > old user size, then extra columns are copies of first column
- if new size < old user size, the resizing is done, but that's all. This means dumping some old TS, but they won't be used anyway, because the TS number are necessarily drawn or chosen in such way they can't be reached.  

Note that for **mingen**, this operation of resizing and filling extra columns is done in the same spirit as it's done right after loading hydro ready made TS.
Eventually (at some point), generating or loading TS should be done at the same place, that is at the place the loading has now, so that : 
- the resizing / filling of mingen can be done once instead of twice today 
-  the  TS mangement is the simplest as possible after that stage of generation/load (no need to manage if TS are generated or loaded, they are just TS)

guilpier-code committed:
Simplify TS numbers draws : repair unit tests

guilpier-code committed:
Simplify TS numbers draws : fixing an error on end-to-end simple study test

guilpier-code committed:
[skip ci] Simplify TS numbers draws : a bit of cleaning

guilpier-code committed:
Simplify TS numbers draws : resetting ts numbers (instead of resizing them)

flomnes commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -716,14 +597,11 @@ void drawAndStoreTSnumbersForNOTintraModal(const array<bool, timeSeriesCount>& i
-        auto& value = group->timeseriesNumbers[0][year];
-        if (nbTimeSeries == 1)
+        auto& groupTsNumber = group->timeseriesNumbers[0][year];
+        groupTsNumber = 0;
+        if (nbTimeSeries != 1)
```
```suggestion
        if (nbTimeSeries == 1)
                groupTsNumber = 0;
        else
                // ...
```

guilpier-code replied:
Now 1-d Matrix **timeseriesNumbers** is reset (it's given zeros everywhere) as soon as it is sized, so we don't need to assign it a zero when associated TS matrix is one column sized.

flomnes commented on src/solver/ts-generator/generator.h:
```diff
@@ -34,12 +34,11 @@
-namespace Solver
-{
-namespace TSGenerator
+namespace Antares::Solver::TSGenerator
```
```suggestion
namespace Antares::TSGenerator
```

guilpier-code replied:
You suggest that namespace **TSGenerator** should not be contained in namespace  **Solver**.
That's right.
I moved  **TSGenerator** from **Antares::Solver** to **Antares** (many changes).

guilpier-code committed:
Simplify TS numbers draws : small simplifications

guilpier-code committed:
Simplify TS numbers draws : testing (on CI) the removal of function fixTSNumbersWhenWidthIsOne

guilpier-code committed:
Simplify TS numbers draws : if hydro TS is generated, mingen should not be modified

guilpier-code committed:
Simplify TS numbers draws : when hydro TS are corrected, mingen size resized just as when TS are ready made

guilpier-code committed:
Simplify TS numbers draws : removing fixTSNumbersWhenWidthIsOne, useless

This function is useless because :
- Why should we modify TS number just before printing them (they are used and then modified when printed. It does not make sense) ?
- the mission is to set TS number to zero when the associated TS is only one, but the default value for TS numbers is now zero, and the TS numbers change
  only if the number of Ts is greater than one.

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/develop' into fix/simplify-ts-number-draws

guilpier-code committed:
Simplify TS numbers draws : create a .cpp file and move non template function into it (fix after code review)

guilpier-code committed:
Simplify TS numbers draws : moving namespace TSGenerator from Antares::Solver to Antares

guilpier-code commented on src/libs/antares/series/series.cpp:
```diff
@@ -74,24 +72,17 @@ int TimeSeries::saveToFile(const std::string& filename, bool saveEvenIfAllZero)
 
 double TimeSeries::getCoefficient(uint32_t year, uint32_t timestep) const
 {
-    if (timeSeries.width == 0)
```
**Explanation** : by construction, **timeSeries** can never have a width == 0, so we don't need the previous complexity

guilpier-code commented on src/libs/antares/series/series.cpp:
```diff
@@ -74,24 +72,17 @@ int TimeSeries::saveToFile(const std::string& filename, bool saveEvenIfAllZero)
 
 uint32_t TimeSeries::getSeriesIndex(uint32_t year) const
 {
-    if (timeSeries.width == 1)
```
**Explanation** : now, at construction of **timeseriesNumbers**, default content is **zero** for every year, that's why we don't need the previous complexity.

guilpier-code commented on src/libs/antares/study/area/area.cpp:
```diff
@@ -277,14 +277,14 @@ void Area::resizeAllTimeseriesNumbers(uint n)
     }
     else
     {
-        load.series.timeseriesNumbers.resize(1, n);
```
Explanation : now we reset (resize + assign **zeros**), which makes TS number **zero** by default. This allows to reduce the amount of code in **timeseries-numbers.cpp** (see changes in **GenerateDeratedMode(...)** for instance)

flomnes replied:
Can put this as a comment in the code ?

guilpier-code replied:
What kind of comment do you want to put exactly ?
I just made some more simplification here, based on the fact that n is a number of years, and cannot be **0**.

guilpier-code replied:
> Can put this as a comment in the code ?

What kind of comment do you want to put exactly ?
I also just made some more simplifications here, based on the fact that **n** is a number of years, and can never be **0** (I checked).

guilpier-code commented on src/tests/src/solver/simulation/tests-ts-numbers.cpp:
```diff
@@ -126,6 +127,7 @@ BOOST_AUTO_TEST_CASE(two_areas_with_5_ready_made_ts_on_load___check_intra_modal_
 	area_2->resizeAllTimeseriesNumbers(1 + study->runtime->rangeLimits.year[rangeEnd]);
 	area_2->load.series.timeSeries.resize(5, 1);
 
+	TSGenerator::ResizeGeneratedTimeSeries(study->areas, study->parameters);
```
**Explanations** : Here we generate all TS we use. Now that we separated TS sizing and TS generation, we must call a function for both.

guilpier-code commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -87,122 +87,56 @@ static bool GenerateDeratedMode(Study& study)
     logs.info() << "  :: using the `derated` mode";
     if (study.parameters.useCustomScenario)
         logs.warning() << "The derated mode is enabled. The custom building mode will be ignored";
-
```
**Explanations** : main changes are in this file.

guilpier-code commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -810,68 +660,6 @@ void applyMatrixDrawsToInterModalModesInArea(Matrix<uint32_t>* tsNumbersMtx,
-    }
-}
-
-static void fixTSNumbersWhenWidthIsOne(Study& study)
```
**Explanations** : now we that we made **zero** a  default value (at construction) for TS numbers, this makes these previous fix functions useless.

guilpier-code commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -722,14 +576,10 @@ void drawAndStoreTSnumbersForNOTintraModal(const array<bool, timeSeriesCount>& i
-        auto& value = group->timeseriesNumbers[0][year];
-        if (nbTimeSeries == 1)
-        {
-            value = 0;
```
We don't need to assign **zero** to a TS number (in case associated size is **1**) : **zero** is the default value now

guilpier-code commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -703,17 +564,10 @@ void drawAndStoreTSnumbersForNOTintraModal(const array<bool, timeSeriesCount>& i
             {
                 auto& link = *(it->second);
                 const uint nbTimeSeries = link.directCapacities.timeSeries.width;
-                if (nbTimeSeries == 1)
```
We don't need to assign **zero** to a TS number (in case associated size is **1**) : **zero** is the default value now

guilpier-code commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -256,63 +176,35 @@ class areaLinksTransCapaNumberOfTSretriever : public areaNumberOfTSretriever
 
-bool IntraModalConsistencyChecker::check()
-{
-    if (isTSintramodal_ && not isTSgenerated_)
```
**Explanations** : we don't have to know if TS are generated or not anymore, that's why this function is removed

guilpier-code commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -912,28 +700,18 @@ bool TimeSeriesNumbers::Generate(Study& study)
     array<uint32_t, timeSeriesCount> intramodal_draws;
     std::fill(intramodal_draws.begin(), intramodal_draws.end(), 0);
 
-    const array<bool, timeSeriesCount> isTSgenerated
```
**Explanations** : same as above : no need to know if TS are generated or not, so no need for this array anymore.

guilpier-code commented on src/libs/antares/study/parts/hydro/series.cpp:
```diff
@@ -236,11 +236,22 @@ void DataSeriesHydro::reset()
-    ror.resize(width, HOURS_PER_YEAR);
-    storage.resize(width, DAYS_PER_YEAR);
-    count = width;
+    ror.resize(newWidth, HOURS_PER_YEAR);
```
**Explanations** : in case we generate hydro TS, **ror** and **storage** are generated, but not **mingen**.
So **ror** and **storage** are resized and completely overridden. But we must be careful with **mingen** TS : it is only resized, but we don't want to loose existing **mingen** TS given by the user : 
- If we extend **mingen** TS, we keep TS given by users (and fill the extra ones with a copy of the first TS)
- if we reduce **mingen** TS, we can remove extra TS given by user, but we must keep the left ones intact

flomnes replied:
Can you include this as a comment ?

guilpier-code replied:
> Can you include this as a comment ?

Done

guilpier-code commented on src/solver/simulation/solver.hxx:
```diff
@@ -312,6 +312,8 @@ void ISimulation<Impl>::run()
 
     ImplementationType::setNbPerformedYearsInParallel(pNbMaxPerformedYearsInParallel);
 
+    TSGenerator::ResizeGeneratedTimeSeries(study.areas, study.parameters);
```
**Explanations** : this is the place where we call the resizing for TS to be generated.

guilpier-code commented on src/solver/ts-generator/hydro.cpp:
```diff
@@ -40,18 +40,12 @@ using namespace Yuni;
-{
 namespace TSGenerator
 {
-static void PreproHydroInitMatrices(Data::Study& study, uint tsCount)
```
**Explanations** : we separated resizing from generation, so we moved this code to another place (see **generator.cpp**)

guilpier-code commented on src/solver/ts-generator/thermal.cpp:
```diff
@@ -251,15 +248,11 @@ void GeneratorTempData::operator()(Data::Area& area, Data::ThermalCluster& clust
 
     if (0 == cluster.unitCount or 0 == cluster.nominalCapacity)
     {
-        cluster.series.timeSeries.reset(1, nbHoursPerYear);
-
```
Explanations : due to separation between TS resizing and generation, this code (**resize**) was moved elsewhere

guilpier-code commented on src/solver/ts-generator/xcast/studydata.hxx:
```diff
@@ -57,18 +55,12 @@ void StudyData::loadFromStudy(Data::Study& study,
             logs.info() << "  Added the area '" << area.name << "'";
             localareas.push_back(&area);
         }
-        else
```
**Explanations** : code moved elsewhere due to separation between TS resizing and generation

guilpier-code committed:
Simplify TS numbers draws : range based for loop instead of iterators (due to code review)

JasonMarechal25 commented on src/solver/ts-generator/generator.hxx:
```diff
@@ -29,12 +29,52 @@
+        for (auto it = area.thermal.list.mapping.begin(); it != end; ++it)
+        {
+            auto& cluster = *(it->second);
+            bool globalThermalTSgeneration = params.timeSeriesToGenerate & Data::timeSeriesThermal;
```
This constant can go outside the loop

guilpier-code replied:
Done 
**Caution** : due to another remark from @sylvlecl, this loop (and the function it belongs to) was moved to a new source file named **generator.cpp**

guilpier-code committed:
[skip ci] Simplify TS numbers draws : range based for loop instead of iterators (due to code review) - part 2

guilpier-code committed:
Simplify TS numbers draws : simplify and clarify function checkAllElementsIdenticalOrOne

payetvin commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -87,122 +87,56 @@ static bool GenerateDeratedMode(Study& study)
+    std::vector<uint> getAreaTimeSeriesNumber(const Area& area) override
     {
-        return study_.parameters.nbTimeSeriesLoad;
+        return { area.load.series.timeSeries.width };
```
Remove braces in return statement

guilpier-code replied:
We want to return a **std::vector**, so we should keep it that way, should we not ?

payetvin commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -87,122 +87,56 @@ static bool GenerateDeratedMode(Study& study)
+    std::vector<uint> getAreaTimeSeriesNumber(const Area& area) override
     {
-        return study_.parameters.nbTimeSeriesHydro;
+        return { area.hydro.series->TScount() };
```
Same here

guilpier-code replied:
Same answer as above

payetvin commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -87,122 +87,56 @@ static bool GenerateDeratedMode(Study& study)
     return true;
 }
 
 class areaNumberOfTSretriever
```
Class names should start with a capital letter by convention

guilpier-code replied:
Done

payetvin commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -87,122 +87,56 @@ static bool GenerateDeratedMode(Study& study)
-    {
-    }
+    virtual ~areaNumberOfTSretriever() = default;
     virtual std::vector<uint> getAreaTimeSeriesNumber(const Area& area) = 0;
```
There's specific cases for hydro and links needing to override the base implementation.

I think the others containers should using the default class and function
The default class should take the series as an argument in constructor.

This would allow to have like
```
class AreaNumberOfTSretriever
{
public:

const TimeSeries& series;

AreaNumberOfTSretriever(const TimeSeries& s) : series(s)
{}

std::vector<uint> getAreaTimeSeriesNumber()
{
    return series.width;
}
};
```

guilpier-code replied:
Yes, you're right !
What you recommend is planned for another PR (about re-thinking this file's design).

payetvin commented on src/solver/simulation/timeseries-numbers.cpp:
```diff
@@ -87,122 +87,56 @@ static bool GenerateDeratedMode(Study& study)
-    {
-    }
-    std::vector<uint> getAreaTimeSeriesNumber(const Area& area)
+    std::vector<uint> getAreaTimeSeriesNumber(const Area& area) override
```
Thermal and renewable behave the same way, using common/cluster_list
We may be able to merge them by passing only clusters

guilpier-code replied:
Sure, you're right, but as you know, there are many changes to come from another PR, for example : **clusterCount()** is about to disappear from **thermal/container.h**, and other things.
To remove the code duplication you recommend, we could wait a bit, so that the part about clusters simplification is done.

payetvin commented on src/solver/ts-generator/xcast/xcast.cpp:
```diff
@@ -43,12 +43,13 @@ using namespace Yuni;
 namespace XCast
 {
+
+const unsigned int nbHoursADay = 24;
```
Maybe define in in constants.h with HOURS_PER_YEAR ?
It's fine declaring it locally so up to you

guilpier-code replied:
Done

flomnes commented on src/libs/antares/study/parts/hydro/series.cpp:
```diff
@@ -236,11 +236,22 @@ void DataSeriesHydro::reset()
+    ror.resize(newWidth, HOURS_PER_YEAR);
+    storage.resize(newWidth, DAYS_PER_YEAR);
+
+    // Resizing mingen : new extra columns are copies of the first column.
+    //    Note that mingen has necessarily one column at least, by construction
```
Why do we want to do that ?

guilpier-code replied:
Please see [this comment](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1752#discussion_r1417403948)

guilpier-code replied:
> Why do we want to do that ?

Please see [this comment](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1752#discussion_r1417403948)

guilpier-code committed:
Simplify TS numbers draws : a bit of cleaning due to a review

guilpier-code committed:
Simplify TS numbers draws : comments and simplifications after code review

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/develop' into fix/simplify-ts-number-draws

guilpier-code committed:
Simplify TS numbers draws : make code compile after merge

guilpier-code committed:
Simplify TS numbers draws : make code slightly more consistent and understandable
